maas/client/viscera/boot_resources.py

```diff
@@ -0,0 +1,285 @@
+            raise ValueError(
+                "name must be in format os/release; missing '/'")
+        if not isinstance(architecture, str):
+            raise TypeError(
+                "architecture must be a str, not %s" % (
+                    type(architecture).__name__))
+        elif '/' not in architecture:
+            raise ValueError(
+                "architecture must be in format arch/subarch; missing '/'")
+        if not isinstance(content, io.IOBase):
+            raise TypeError(
+                "content must extend from io.IOBase; %s does not" % (
+                    type(content).__name__))
+        elif not content.readable():
+            raise ValueError("content must be readable")
+        elif not content.seekable():
+            raise ValueError("content must be seekable")
+        if title is None:
+            title = ""
+        if not isinstance(title, str):
+            raise TypeError(
+                "title must be a str, not %s" % type(title).__name__)
+        if not isinstance(filetype, BootResourceFiletype):
+            raise TypeError(
+                "filetype must be a BootResourceFiletype, not %s" % (
+                    type(filetype).__name__))
+        if not isinstance(chunk_size, int):
+            raise TypeError(
+                "chunk_size must be a int, not %s" % (
+                    type(chunk_size).__name__))
+        elif chunk_size <= 0:
+            raise ValueError(
+                "chunk_size must be greater than 0, not %d" % chunk_size)
+        if (progress_callback is not None and
+                not isinstance(progress_callback, Callable)):
+            raise TypeError(
+                "progress_callback must be a Callable, not %s" % (
+                type(progress_callback).__name__))
+
+        size, sha256 = calc_size_and_sha265(content, chunk_size)
+        resource = cls._object(cls._handler.create(
+            name=name, architecture=architecture, title=title,
+            filetype=filetype.value, size=str(size), sha256=sha256))
+        newest_set = sorted(resource.sets.keys(), reverse=True)[0]
+        resource_set = resource.sets[newest_set]
+        rfile = list(resource_set.files.values())[0]
+        if rfile.complete:
+            # Already created and fully up-to-date.
+            return resource
+        else:
+            # Upload in chunks and reload boot resource.
+            cls._upload_chunks(rfile, content, chunk_size, progress_callback)
+            return cls._object.read(resource.id)
+
+    @classmethod
+    def _upload_chunks(
+            cls, rfile: BootResourceFile, content: io.IOBase, chunk_size: int,
+            progress_callback: Callable=None):
+        """Upload the `content` to `rfile` in chunks using `chunk_size`."""
+        content.seek(0, io.SEEK_SET)
+        upload_uri = "%s%s" % (
+            cls._handler.uri[:len(cls._handler.uri)-len(cls._handler.path)],
+            rfile._data['upload_uri'])
```

Comment:
Okay, but abstractions are leaky. You could make it an `_upload_uri` attribute of the object instead.